mlir/include/mlir/IR/Operation.h

```diff
@@ -1019,16 +1019,12 @@ class alignas(8) Operation final
 
   /// Expose a few methods explicitly for the debugger to call for
   /// visualization.
-#ifndef NDEBUG
-  LLVM_DUMP_METHOD operand_range debug_getOperands() { return getOperands(); }
-  LLVM_DUMP_METHOD result_range debug_getResults() { return getResults(); }
-  LLVM_DUMP_METHOD SuccessorRange debug_getSuccessors() {
-    return getSuccessors();
-  }
-  LLVM_DUMP_METHOD MutableArrayRef<Region> debug_getRegions() {
-    return getRegions();
-  }
-#endif
+  /// @{
+  operand_range debug_getOperands();
```

Comment:
This makes it possible to turn a compile-time failure into a link-time failure, that seems like a "regression" to me in behavior. 

You mention a problem of flag divergence, however that's what llvm-config.h is here to resolve.

And actually there is a FIXME before the definition of `LLVM_DUMP_METHOD` to this effect:

> // FIXME: Move this to a private config.h as it's not usable in public headers.

Reply:
So the FIXME was fixed by llvm-config.h?

However, the `LLVM_DUMP_METHOD` makes use of the `NDEBUG` macro which is under the including's application's control and `LLVM_ENABLE_DEBUG` is not present in `llvm-config.h`.


> This makes it possible to turn a compile-time failure into a link-time failure, that seems like a "regression" to me in behavior.

This is by design of the `LLVM_DUMP_METHOD` pattern established in 8c209aa8779de57771b64896f805e14cc0016dcd. You were not supposed to call `debug_getOperands()` in code, just from the debugger, so if you had a compile-time failure, you were using it wrong. I think there wouldn't be a lot of bloat if we also had dump functions in release builds, but I think such a change should be LLVM-wise.

I guess one could make some dump functions be including-application controlled by user application macros. However, the cost is high: because by design there is are no uses of such functions, `__attribute__((__used__))` must be used which causes it to be emitted in every single translation unit. 

Anyway, there should be a correct use of the `LLVM_DUMP_METHOD` macro, and MLIR currently doesn't follow any pattern.

Reply:
> So the FIXME was fixed by llvm-config.h? However, the LLVM_DUMP_METHOD makes use of the NDEBUG macro which is under the including application's control and LLVM_ENABLE_DEBUG is not present in llvm-config.h.

No it's not fixed, it should be fixed by making use of llvm-config.h to not rely on NDEBUG.

> This is by design of the LLVM_DUMP_METHOD pattern established in https://github.com/llvm/llvm-project/commit/8c209aa8779de57771b64896f805e14cc0016dcd.

I don't think the aspect of "not failing at compile time" is actually desirable "by design" here, why would it be better than relying on llvm-config.h ?

Reply:
It's not about desirability, but following LLVM coding practice.

That can be changed of course, but will require changes through all of LLVM (due to changing the `LLVM_DUMP_METHOD` macro) and possibly require a community discussion. In the meantime, this would be the fix for making most `dump()` methods actually callable in the debugger. I'd add the suggested usability improvement to not declare those functions depending on `llvm-config.h` in a followup.

Reply:
Sorry, I don't agree that this is a "coding practice" just because this is a pattern appearing somewhere else.
In particular the existing FIXME in the code acknowledge that this isn't how it should be setup, and I don't understand why changing this as you do right now is a good thing to do, instead of addressing the FIXME.